internal/metric/metric.go

```diff
@@ -17,133 +17,37 @@ limitations under the License.
 package metric
 
 import (
+	"context"
 	"fmt"
+	"math"
 	"strconv"
-	"strings"
-	"time"
 
 	"github.com/go-logr/logr"
 	redskyv1beta1 "github.com/redskyops/redskyops-controller/api/v1beta1"
 	"github.com/redskyops/redskyops-controller/internal/template"
-	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-// CaptureError describes problems that arise while capturing metric values
-type CaptureError struct {
-	// A description of what went wrong
-	Message string
-	// The URL that was used to capture the metric
-	Address string
-	// The metric query that failed
-	Query string
-	// The completion time at which the query was executed
-	CompletionTime time.Time
-	// The minimum amount of time until the metric is expected to be available
-	RetryAfter time.Duration
-}
-
-func (e *CaptureError) Error() string {
-	return e.Message
-}
-
-// CaptureMetric captures a point-in-time metric value and it's error (standard deviation)
-func CaptureMetric(log logr.Logger, metric *redskyv1beta1.Metric, trial *redskyv1beta1.Trial, target runtime.Object) (float64, float64, error) {
-	// Work on a copy so we can render the queries in place
-	metric = metric.DeepCopy()
-
-	// Execute the query as a template against the current state of the trial
+// CaptureMetric captures a point-in-time metric value and it's error rate.
+func CaptureMetric(ctx context.Context, log logr.Logger, trial *redskyv1beta1.Trial, metric *redskyv1beta1.Metric, target runtime.Object) (float64, float64, error) {
+	// Execute the queries as Go templates
 	var err error
 	if metric.Query, metric.ErrorQuery, err = template.New().RenderMetricQueries(metric, trial, target); err != nil {
 		return 0, 0, err
 	}
 
 	// Capture the value based on the metric type
 	switch metric.Type {
-	case redskyv1beta1.MetricLocal, redskyv1beta1.MetricPods, "":
-		// Just parse the query as a float
+	case redskyv1beta1.MetricKubernetes, "":
 		value, err := strconv.ParseFloat(metric.Query, 64)
-		return value, 0, err
+		return value, math.NaN(), err
 	case redskyv1beta1.MetricPrometheus:
-		return capturePrometheusMetric(log, metric, target, trial.Status.CompletionTime.Time)
+		return capturePrometheusMetric(ctx, log, metric, trial.Status.CompletionTime.Time)
 	case redskyv1beta1.MetricDatadog:
-		return captureDatadogMetric(metric.Scheme, metric.Query, trial.Status.StartTime.Time, trial.Status.CompletionTime.Time)
+		return captureDatadogMetric(metric, trial.Status.StartTime.Time, trial.Status.CompletionTime.Time)
 	case redskyv1beta1.MetricJSONPath:
-		return captureJSONPathMetric(metric, target)
+		return captureJSONPathMetric(metric)
```

Comment:
this seems to surface the pattern of a metric target/collector being `func(ctx, log, metric, start time, end time)`; might be a good candidate for an interface.